src/server/main_service.cc

```diff
@@ -1326,6 +1325,196 @@ bool StartMultiExec(DbIndex dbid, Transaction* trans, ConnectionState::ExecInfo*
   return true;
 }
 
+template <typename F> void IterateKeys(CmdArgList args, KeyIndex keys, F&& f) {
+  for (unsigned i = keys.start; i < keys.end; i += keys.step)
+    f(args[i]);
+
+  if (keys.bonus)
+    f(args[keys.bonus]);
+}
+
+// ExecutionSquasher allows executing a series of commands under a multi transaction
+// and squashing multiple consecutive single-shard commands into one hop whenever it's possible,
+// thus greatly decreasing the dispatch overhead for them.
+class ExecutionSquasher {
+ public:
+  static void Execute(absl::Span<StoredCmd> cmds, ConnectionContext* cntx) {
+    ExecutionSquasher{cmds, cntx}.Run();
+  }
+
+ private:
+  using ReplyChan = ::util::fibers_ext::SimpleChannel<CapturingReplyBuilder::Payload>;
+
+  // Per-shard exection info.
+  struct ShardExecInfo {
+    ShardExecInfo() : cmds{}, reply_chan{nullptr}, local_tx{nullptr} {
+    }
+
+    std::vector<StoredCmd*> cmds;  // accumulated commands
+    unique_ptr<ReplyChan> reply_chan;
+    intrusive_ptr<Transaction> local_tx;  // stub-mode tx for use inside shard
+  };
+
+  static constexpr int kChanBufferSize = 16;
+
+ private:
+  ExecutionSquasher(absl::Span<StoredCmd> cmds, ConnectionContext* cntx);
+
+  // Lazy initialize shard info
+  ShardExecInfo& PrepareShardInfo(ShardId sid);
+
+  // Return true if command was squashed.
+  bool TrySquash(StoredCmd* cmd);
+
+  // Execute separate non-squashed cmd.
+  void ExecuteStandalone(StoredCmd* cmd);
+
+  // Callback that runs on shards during squashed hop.
+  OpStatus SquashedHopCb(Transaction* parent_tx, EngineShard* es);
+
+  // Execute all currently squashed commands.
+  void ExecuteSquashed();
+
+  // Run all commands until completion.
+  void Run();
+
+ private:
+  absl::Span<StoredCmd> cmds_;  // Input range of stored commands
+  ConnectionContext* cntx_;     // Underlying context
+  const CommandId* base_cid_;   // either EVAL or EXEC, used for squashed hops
+
+  std::vector<ShardExecInfo> sharded_;
+  std::vector<ShardId> order_;  // reply order for squashed cmds
+
+  // multi modes that lock on hops (non-atomic, incremental) need keys for squashed hops.
+  // track_keys_ stores whether to populate collected_keys_
+  bool track_keys_;
+  absl::flat_hash_set<MutableSlice> collected_keys_;
```

Comment:
Transaction filters duplicated keys out internally either way, so a vector would be enough for tracking, however we maybe still want to track how many _unique_ keys we are going to run on - and decide if its  "too many". Because for non-atomic running in large batches can actually create lots of contentions just because of locking

Reply:
I however don't have a good heuristic yet - and I'm not sure if its possible at all to have a good one. For example, only a single heavily contented key (like a counter, etc...) would seriously slow down squashed execution (maybe we can use some of the top-keys stuff in the future for evaluation?)